query/src/sql/planner/binder/ddl/table.rs

```diff
@@ -369,8 +372,32 @@ impl<'a> Binder {
                     .collect();
                 (DataSchemaRefExt::create(fields), vec![])
             }
-            // TODO(leiysky): Support `CREATE TABLE AS SELECT` with specified column definitions
-            _ => Err(ErrorCode::UnImplement("Unsupported CREATE TABLE statement"))?,
+            (Some(source), Some(query)) => {
+                // e.g. `CREATE TABLE t (i INT) AS SELECT * from old_t` with columns speicified
+                let (source_schema, source_coments) =
+                    self.analyze_create_table_schema(source).await?;
+                let init_bind_context = BindContext::new();
+                let (_s_expr, bind_context) = self.bind_query(&init_bind_context, query).await?;
+                let query_fields: Vec<DataField> = bind_context
+                    .columns
+                    .iter()
+                    .map(|column_binding| {
+                        DataField::new(
+                            &column_binding.column_name,
+                            *column_binding.data_type.clone(),
+                        )
+                    })
+                    .collect();
+                let  source_fields = source_schema.fields().clone();
+                let source_fields = self.concat_fields(source_fields, query_fields);
+                (
+                    DataSchemaRefExt::create(source_fields.to_vec()),
+                    source_coments,
+                )
+            }
```

Comment:
I found the semantic of `CREATE TABLE AS` with column definitions is wierd, and both `MySQL` and `Postgres` doesn't support it. 

I suggest to raise any error for such case.

Reply:
What do you mean?
 I run the create table with specified columns in mysql. It supported that. 
```sql
mysql> create table t1(i INT, j Int);
Query OK, 0 rows affected (0.01 sec)

mysql> insert into t1 values (1, 2);
Query OK, 1 row affected (0.00 sec)

mysql> create table t2 (k Int, l Int) as select * from t1;
Query OK, 1 row affected (0.00 sec)
Records: 1  Duplicates: 0  Warnings: 0

mysql> desc t2;
+-------+------+------+-----+---------+-------+
| Field | Type | Null | Key | Default | Extra |
+-------+------+------+-----+---------+-------+
| k     | int  | YES  |     | NULL    |       |
| l     | int  | YES  |     | NULL    |       |
| i     | int  | YES  |     | NULL    |       |
| j     | int  | YES  |     | NULL    |       |
+-------+------+------+-----+---------+-------+
4 rows in set (0.00 sec)
mysql> select version();
+-----------+
| version() |
+-----------+
| 8.0.21    |
+-----------+
1 row in set (0.02 sec)
```

Reply:
I was wrong, MySQL does support this. 

If the data types of coulumn definition is incompatible with the select result, we may need to raise an error for that.

Reply:
If the schema is not matched, there will be a `TransformCastSchema` between them.

This is some as:  `insert table t select * from t2`

Reply:
But I found that it is needed to check the cast type here. Otherwise, the table has been created, even though the server throws an error from the `InsertInterpreter`. :)

Reply:
> But I found that it is needed to check the cast type here. Otherwise, the table has been created, even though the server throws an error from the `InsertInterpreter`. :)

By the way, we have a plan for atomic DDL, which is tracked in #6625. You can take a took if you are interested in it.